### DISCUSSION TITLE
Can I use Lexical with Remix?

### DISCUSSION THREAD

**Discussion by rodolphoasb** [remix-run/remix#7623] [Q&A] [answered]
Hey, I stumbled upon [this issue](https://github.com/facebook/lexical/issues/1177) in the lexical repo and was wondering if this changes with v2.

Is it possible to use Lexical with Remix?

**Comment by kiliman** [comment, marked as answer]
Typically with components that don't really support SSR, you would wrap that in a `<ClientOnly>` component from `remix-utils`

First create a _lexical.client.tsx_ file and re-export all the lexical functions
```ts
// components/lexical.client.tsx
export * from 'lexical'
export * from '@lexical/react'
```
Then in your route, import `ClientOnly` and your lexical components
```ts
// routes/editor.tsx
import { ClientOnly } from 'remix-utils'
import { LexicalComposer } from '~/components/lexical.client'

export default Component() {
  return (
    <ClientOnly fallback={<p>Loading</p>}>
      {() => (
        <LexicalComposer>
          //...
        </LexicalComposer>
      )}
    </ClientOnly>
  )
}
```

**Comment by luisrudge** [comment]
@rodolphoasb did you figure this one out? @kiliman's solution doesn't work because there's no export from `@lexical/react`

**Reply by rodolphoasb** [reply]
Hey, @luisrudge! Yeah you can use it directly now. Related: https://github.com/facebook/lexical/issues/5120. Make sure you are using the lastest Lexical version.

**Reply by luisrudge** [reply]
Thanks - I guess my issue is not the same then 😅
My issue is more with all the plugins in the playground not being compatible with SSR. I wanted to basically copy/paste the playground into a remix app and run it, but it doesn't work. lots of plugins trying to access `window` or `document`. So frustrating 😵‍💫

**Reply by rodolphoasb** [reply]
@luisrudge Oh, in that case I think I have the solution for you!

Try it this way inside the `Editor`:

```
import { ClientOnly } from "remix-utils/client-only";

<ClientOnly>{() => <ComponentPickerPlugin />}</ClientOnly>
```

Btw, you only need to do this with those plugins that you mentioned. There's a few of them that work without it.

**Reply by luisrudge** [reply]
oh.. you mean replacing all the plugin imports with clientonly? hmm.. that might work. I didn't try that, but I tried

```tsx
<ClientOnly>
  {()=><Editor/>}
</ClientOnly>
```
and it didn't work because the code that throws the error runs when it's imported, not when it's run.

example: https://github.com/facebook/lexical/blob/main/packages/lexical-playground/src/plugins/ImagesPlugin/index.tsx#L262

**Reply by rodolphoasb** [reply]
Hope this helps: https://cap.link/3pbwz9jdjhtf4zy

**Reply by luisrudge** [reply]
Thanks, I appreciate the help! I'll try this if I need one of those plugins!

**Reply by luisrudge** [reply]
I tested this approach and it still doesn't work

```
//imports
import { ClientOnly } from 'remix-utils/client-only';
import ComponentPickerPlugin from './plugins/ComponentPickerPlugin';


//rendering
<AutoFocusPlugin />
<ClearEditorPlugin />
<ClientOnly>{() => <ComponentPickerPlugin />}</ClientOnly>
<EmojiPickerPlugin />
<AutoEmbedPlugin />
```

error:
<img width="1585" alt="image" src="https://github.com/user-attachments/assets/b7bda642-7c4f-4dc3-b7f3-77f604649465">


seems like the LexicalTypeaheadMenuPlugin import is already reading from `document`, which is throwing right away

**Reply by luisrudge** [reply]
oh, duh. I had to wrap other plugins in the <ClientOnly> 😅
it works now. do you also get a bunch of log messages like this?
<img width="1177" alt="image" src="https://github.com/user-attachments/assets/206dc0d9-6af9-4b04-a3ee-296d4eb7a7d0">
